### PR TITLE
feat(extensions): support .extensionignore to exclude files during install

### PR DESCRIPTION
**Pull request by Rubiss**
## Summary

Adds `.extensionignore` support so extension authors can exclude files and folders from being copied when users run `specify extension add`.

## Problem

Extension authors often have development-only files (tests, CI configs, documentation source, build artifacts) that shouldn't be copied into the user's project when installing an extension. There was no mechanism to exclude these files.

## Solution

Extension authors can now create a `.extensionignore` file in their extension root. The file uses `.gitignore`-compatible patterns (one per line), powered by the [`pathspec`](https://pypi.org/project/pathspec/) library:

```gitignore
# .extensionignore

# Development files
tests/
.github/
.gitignore

# Build artifacts
__pycache__/
*.pyc
```

### Features
- `.gitignore`-compatible pattern matching via `pathspec.GitIgnoreSpec`
- `*` matches anything except `/`; `**` matches zero or more directories
- `?` matches any single character except `/`
- Trailing `/` on directory patterns restricts matching to directories only
- Patterns containing `/` (other than trailing) are anchored to the extension root
- `!` negates a previously excluded pattern (re-includes a file)
- Comment lines (`#`) and blank lines are ignored
- Backslash patterns are normalised to forward slashes for cross-platform compatibility
- The `.extensionignore` file itself is always excluded from the copy

## Changes

- **src/specify_cli/extensions.py**: Added `_load_extensionignore()` static method and integrated it into `install_from_directory()` via `shutil.copytree(ignore=...)`
- **extensions/EXTENSION-DEVELOPMENT-GUIDE.md**: New documentation section covering `.extensionignore` format, example patterns, and a pattern-matching reference table
- **tests/test_extensions.py**: 13 new tests in `TestExtensionIgnore` covering all pattern matching scenarios
- **pyproject.toml**: Added `pathspec>=0.12.0` dependency
- **CHANGELOG.md**: Added `[Unreleased]` entry for this feature

## Testing

All 152 passing tests continue to pass (4 pre-existing failures in `test_cursor_frontmatter.py` are unrelated — bash/WSL not available on Windows). New tests cover:

- No `.extensionignore` → all files copied (baseline)
- Directory exclusion (`tests/`, `.github/`)
- Glob patterns (`*.pyc`) across nested directories
- Comments and blank lines ignored
- `.extensionignore` itself never copied
- Relative path pattern matching
- `..` traversal patterns are a no-op (security)
- Absolute path patterns are a no-op (security)
- Empty `.extensionignore` file
- Windows backslash patterns normalised correctly
- `*` does not cross directory boundaries
- `**` crosses directory boundaries
- `!` negation re-includes previously excluded files

## AI Disclosure

This PR was authored with GitHub Copilot assistance.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,12 @@ Recent changes to the Specify CLI and templates are documented here.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.1.14] - 2026-03-09
+
+### Added
+
+- feat(extensions): support `.extensionignore` to exclude files/folders during `specify extension add` (#TBD)
```
[the remark this answers is not in the document]
Comment (Rubiss) [reply CHANGELOG.md]: Done

**Comment by Rubiss** [comment]
> I will approve, but consider the things below
> 
> 1. **`fnmatch` ≠ `.gitignore` semantics (documentation bug)**  
>    The docs and PR description claim "similar to `.gitignore` syntax," but `fnmatch.fnmatch()` treats `*` as matching **any** character including path separators. In `.gitignore`, `*` does **not** cross directory boundaries. The documentation table says `docs/*.draft.md` matches "Draft markdown files inside `docs/`", but `fnmatch` would also match `docs/sub/api.draft.md`. This needs to either:
>    - Switch to `pathlib.PurePath.match()` or `fnmatch` with per-segment matching to get proper gitignore-like behavior, **or**
>    - Remove all `.gitignore` comparisons from the docs and clearly document that `*` matches across directories.
> 
> 2. **No support for negation (`!`) or `**` patterns**  
>    Not a blocker since these aren't claimed, but the `.gitignore` analogy sets expectations. At minimum, the docs should state what's **not** supported.
> 
> 3. **Missing edge-case tests:**
>    - Pattern with `..` (e.g., `../sibling/`) — should be a no-op or explicitly rejected
>    - Absolute path patterns (e.g., `/etc/passwd`) — verify they don't match anything
>    - Empty `.extensionignore` file (no patterns, just the file itself)
>    - Patterns with Windows-style backslashes
> 
> 4. **Minor: forward-slash normalization is one-directional**
>    The code does `rel_path_fwd = rel_path.replace("\\", "/")` but doesn't normalize the patterns themselves. A user on Windows writing `docs\internal\draft.md` in their `.extensionignore` would get inconsistent behavior.

No need to approve yet, these seem like all valid concerns that I will address. Thanks for the feedback. Personally, I think it should work consistently like .gitignore does since generally people are very familiar with the concept.

**Comment by Rubiss** [comment]
> I will approve, but consider the things below
> 
>     1. **`fnmatch` ≠ `.gitignore` semantics (documentation bug)**
>        The docs and PR description claim "similar to `.gitignore` syntax," but `fnmatch.fnmatch()` treats `*` as matching **any** character including path separators. In `.gitignore`, `*` does **not** cross directory boundaries. The documentation table says `docs/*.draft.md` matches "Draft markdown files inside `docs/`", but `fnmatch` would also match `docs/sub/api.draft.md`. This needs to either:
>        
>        * Switch to `pathlib.PurePath.match()` or `fnmatch` with per-segment matching to get proper gitignore-like behavior, **or**
>        * Remove all `.gitignore` comparisons from the docs and clearly document that `*` matches across directories.
> 
>     2. **No support for negation (`!`) or `**` patterns**
>        Not a blocker since these aren't claimed, but the `.gitignore` analogy sets expectations. At minimum, the docs should state what's **not** supported.
> 
>     3. **Missing edge-case tests:**
>        
>        * Pattern with `..` (e.g., `../sibling/`) — should be a no-op or explicitly rejected
>        * Absolute path patterns (e.g., `/etc/passwd`) — verify they don't match anything
>        * Empty `.extensionignore` file (no patterns, just the file itself)
>        * Patterns with Windows-style backslashes
> 
>     4. **Minor: forward-slash normalization is one-directional**
>        The code does `rel_path_fwd = rel_path.replace("\\", "/")` but doesn't normalize the patterns themselves. A user on Windows writing `docs\internal\draft.md` in their `.extensionignore` would get inconsistent behavior.

Thanks for the feedback. I have updated .extensionignore implementation to match familiar .gitignore patterns with a callout of what is not supported. Let me know what you think.

File: pyproject.toml [hunk 2]
```diff
@@ -13,6 +13,7 @@ dependencies = [
     "truststore>=0.10.4",
     "pyyaml>=6.0",
     "packaging>=23.0",
+    "pathspec>=0.12.0",
```
[the remark this answers is not in the document]
Comment (Rubiss) [reply pyproject.toml:16]: Updated the PR description.

File: extensions/EXTENSION-DEVELOPMENT-GUIDE.md [hunk 3]
```diff
@@ -332,6 +332,66 @@ echo "$config"
...
+
+- Blank lines are ignored
+- Lines starting with `#` are comments
+- `*` matches anything **except** `/` (does not cross directory boundaries)
+- `**` matches zero or more directories (e.g., `docs/**/*.draft.md`)
+- `?` matches any single character except `/`
+- A trailing `/` restricts a pattern to directories only
+- Patterns containing `/` (other than a trailing slash) are anchored to the extension root
+- Patterns without `/` match at any depth in the tree
+- `!` negates a previously excluded pattern (re-includes a file)
```
[the remark this answers is not in the document]
Comment (Rubiss) [reply extensions/EXTENSION-DEVELOPMENT-GUIDE.md:351]: Added this to the unsupported features list

**Review by mnriem — APPROVED** [review]

**Comment by mnriem** [comment]
@Rubiss Thank you!
